### PR TITLE
Fix Tracking Layer

### PR DESCRIPTION
**Pull request by stefanhahmann**
The cluster layer was not generated for layers of type tracking layer in function generate_cluster_tracks()

This was caused by bugs in 2 functions():
* The label ids were wrongly taken from the cluster id column instead of the "labels" column in generate_cluster_tracks()
* The generate_cluster_image() function did not map the label ids to the cluster ids in the case of tracking layers

This PR adds a bugfix for this. In order to do so also some recent changes regarding performance optimization had to be reverted (https://github.com/BiAPoL/napari-clusters-plotter/commit/1c2ec0f3669130721fcaab85293ef49eaaf0dc62, https://github.com/BiAPoL/napari-clusters-plotter/commit/dcd83c1ad5842d57506513f8139f28c531a85740, https://github.com/BiAPoL/napari-clusters-plotter/commit/5fee912a3a2dc71fda1c7e2d0b5271174eb2aeea, https://github.com/BiAPoL/napari-clusters-plotter/commit/0800f48b355124f57c4533d68e7440cb002f5250)

@thorstenwagner there is now a unit test for the cluster plotter in combination with tracking layers. Perhaps you could try, if you could make your performance optimizations with this as well?

The PR additionally adds some unit tests to check for bugs that had recently been introduced.

### CODE REVIEW

File: napari_clusters_plotter/_tests/test_utilities.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    assert mapped_image_2d[2, 1] == 1
+    assert mapped_image_2d[2, 2] == 2
+
+
+def test_map_array_4d():
+    from skimage.util import map_array
+
+    output_array = map_array(label_image_4d, input_values, output_values)
+
+    assert output_array[0, 0, 0, 0] == 0
```
[the remark this answers is not in the document]
Comment (stefanhahmann) [reply napari_clusters_plotter/_tests/test_utilities.py]: Thanks for the advice. I changed the code accordingly.

**Comment by Cryaaa** [comment]
Okay so some testing on the very simple tracking example you implemented in the unit test showed me that actually the problem did not lie with the generate_cluster_immage function but indeed was caused by the wrong extraction of labels in the generate_cluster_tracks function. 



> * The generate_cluster_image() function did not map the label ids to the cluster ids in the case of tracking layers

Regarding this: do you maybe have a simple tracking example dataset you might want to share which I can use for testing? After I have tested this in detail I will adjust the code and merge. Thanks for the unit tests again!

**Comment by thorstenwagner** [comment]
Hm, I'm a bit confused. I just checkout your branch, put my new `generate_cluster_image` back in and all tests passed.

```
collecting ... collected 4 items

../../../../../../../../../home/twagner/Projects/napari-clusters-plotter/src/napari-clusters-plotter/napari_clusters_plotter/_tests/test_utilities.py::test_map_array_3d 
../../../../../../../../../home/twagner/Projects/napari-clusters-plotter/src/napari-clusters-plotter/napari_clusters_plotter/_tests/test_utilities.py::test_map_array_4d 
../../../../../../../../../home/twagner/Projects/napari-clusters-plotter/src/napari-clusters-plotter/napari_clusters_plotter/_tests/test_utilities.py::test_generate_cluster_image 
../../../../../../../../../home/twagner/Projects/napari-clusters-plotter/src/napari-clusters-plotter/napari_clusters_plotter/_tests/test_utilities.py::test_generate_cluster_tracks 

======================== 4 passed, 3 warnings in 1.48s =========================
```

**Comment by Cryaaa** [comment]
@stefanhahmann,
I have changed this PR a bit now, the generate cluster image function will be kept as the updated one from @thorstenwagner but the other bug fix will be included. I added your new tests to the test_utils.py file to have everything in one place and have removed duplicate tests and test cases that do not apply to the plotter (e.g. just testing the generate cluster image function on 4D data). I would love to hear if this now works for your data!

File: napari_clusters_plotter/_utilities.py [hunk 2]
```diff
@@ -330,7 +330,7 @@ def generate_cluster_image(label_image, label_list, predictionlist):
 
     # label_list can be removed from method.
     predictionlist_new = np.array(predictionlist) + 1
-    plist = np.zeros(int(np.max(label_image)) + 1, dtype=np.uint32)
+    plist = np.zeros(len(predictionlist) + 1, dtype=np.uint32)
```
Comment (thorstenwagner) [inline napari_clusters_plotter/_utilities.py:333]: Could comment why you think this is equivalent/correct?

**Comment by stefanhahmann** [comment]
> Hm, I'm a bit confused. I just checkout your branch, put my new `generate_cluster_image` back in and all tests passed.

Sorry for the confusion. I updated the respective unit test so that it actually fails, cf.: https://github.com/BiAPoL/napari-clusters-plotter/pull/286/commits/1d952d75f2fd288b8687a30803596084fc750144

The unit test did pass, since both timeframes included in it had all labels of the complete timelapse. If a single timeframe does not have all timeframes of the complete timelapse, it causes an array index out of bounds error.

@Cryaaa and @thorstenwagner in https://github.com/BiAPoL/napari-clusters-plotter/pull/286/commits/b6e5f60a5c6134f54c08e01b862742dbb2cd937e I have committed a suggestion how to fix this error with the new generate_cluster_image() version, but this breaks other tests: https://github.com/BiAPoL/napari-clusters-plotter/actions/runs/7212283328/job/19649645462 (test_cluster_image_generation_unsorted_non_sequential_labels() and test_generate_cluster_4d_labels())

Only with generate_cluster_image_old() all tests pass at the moment, cf.: https://github.com/BiAPoL/napari-clusters-plotter/pull/286/commits/595797719dcfda91f34a1dc399b16356e644c2fc and https://github.com/BiAPoL/napari-clusters-plotter/actions/runs/7212550935/job/19650463473

Currently, I have no good idea how to make the new version of generate_cluster_image() pass all tests.

**Comment by thorstenwagner** [comment]
With my fix, all tests pass. 

The problem is that my method assumes that label_list only contains labels that are actually part of the label_image, which is not the case for your track data. I think it should, and the implementation of `generate_cluster_tracks` may be buggy as it produces the same label_list for each time point, regardless of the actual labels at that time point. However, I have not had a close look yet.

**Comment by thorstenwagner** [comment]
Regarding my last commit, its indeed almost twice as fast:

image

**Comment by thorstenwagner** [comment]
Regarding my very last commit of today: With that we reduce the it even to

 `4.42 µs ± 16 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)`

However, this is probably not relevant for the complete runtime.

**Comment by thorstenwagner** [comment]
I just want to mention: As soon [this](https://github.com/scikit-image/scikit-image/pull/7266#) is merged, we can probably go back to the 'old' map_array function :-)

**Comment by stefanhahmann** [comment]
> With my fix, all tests pass.

Thanks for the fix.

> The problem is that my method assumes that label_list only contains labels that are actually part of the label_image, which is not the case for your track data. 

Thanks for clarifying the assumptions of the new generate_cluster_image() method. 

> I think it should, and the implementation of `generate_cluster_tracks` may be buggy as it produces the same label_list for each time point, regardless of the actual labels at that time point. However, I have not had a close look yet.

With the old implementation it did not matter, which labels were contained at which timepoint. However, I agree, that including only those labels that exist in the considered timepoint should be included. generate_cluster_tracks() can be optimized in this regard. Thanks to you fix, functionality does not currently depend on this optimization.

**Comment by Cryaaa** [comment]
@thorstenwagner, @stefanhahmann: I think we can merge this now that you have fixed some of the more specific details?

**Comment by thorstenwagner** [comment]
Lgtm, from my side you can merge.

**Pull request merged**